### PR TITLE
Adds WASI support to markdown rust example.

### PR DESCRIPTION
**Pull request by yurydelendik**
The wasm files + interface types can be compiled for wasm32-wasi target (e.g. via cargo wasi). That can add some "wasi_unstable" dependecies. In most of the cases it is panic error printing or random number initialization.

This PR allows markdown example to be executed even if wasm file contains "wasi_unstable" imports.

### CODE REVIEW

File: wasmtime-interface-types/src/lib.rs [hunk 1]
```diff
@@ -100,6 +100,20 @@ impl ModuleData {
         })
     }
 
+    /// Detects if WASI support is needed: returns module name that is requested.
+    pub fn has_wasi(&self) -> Option<String> {
```
Comment (alexcrichton) [inline wasmtime-interface-types/src/lib.rs]: Perhaps something like `find_wasi_module_name`?

File: misc/wasmtime-rust/macro/src/lib.rs [hunk 2]
```diff
@@ -64,6 +64,16 @@ fn generate_load(item: &syn::ItemTrait) -> syn::Result<TokenStream> {
...
             });
             let data = #root::wasmtime_interface_types::ModuleData::new(&bytes)?;
+            if let Some(wasi_name) = data.has_wasi() {
+                let wasi_handle = wasmtime_wasi::instantiate_wasi(
+                    "",
+                    cx.get_global_exports(),
+                    &[],
+                    &[],
+                    &[],
+                ).expect("wasi support");
```
Comment (alexcrichton) [inline misc/wasmtime-rust/macro/src/lib.rs]: Could this use `?` to propagate the error?

File: misc/wasmtime-rust/macro/src/lib.rs [hunk 3]
```diff
@@ -64,6 +64,16 @@ fn generate_load(item: &syn::ItemTrait) -> syn::Result<TokenStream> {
                 ..Default::default()
             });
             let data = #root::wasmtime_interface_types::ModuleData::new(&bytes)?;
+            if let Some(wasi_name) = data.has_wasi() {
+                let wasi_handle = wasmtime_wasi::instantiate_wasi(
+                    "",
```
Comment (alexcrichton) [inline misc/wasmtime-rust/macro/src/lib.rs]: Should this string be `wasi_name`?

I'm not actually sure what this string is...

Comment (yurydelendik) [reply misc/wasmtime-rust/macro/src/lib.rs]: using `module_name` here

File: wasmtime-interface-types/src/lib.rs [hunk 4]
```diff
@@ -100,6 +100,20 @@ impl ModuleData {
...
     }
 
+    /// Detects if WASI support is needed: returns module name that is requested.
+    pub fn find_wasi_module_name(&self) -> Option<String> {
+        self.inner.as_ref().and_then(|Inner { module }| {
+            module
+                .imports
+                .iter()
+                .find(|walrus::Import { module, .. }| match module.as_str() {
+                    "wasi" | "wasi_unstable" => true,
```
Comment (sunfishcode) [inline wasmtime-interface-types/src/lib.rs]: WASI will have its first new snapshot soon, which will contain modules named "wasi_snapshot_*".

Comment (jlb6740) [reply wasmtime-interface-types/src/lib.rs]: Is wasi and wasi_unstable needed here after the snapshot? Is it intended for all modules to be wasi_unstable_<something>?

Comment (yurydelendik) [reply wasmtime-interface-types/src/lib.rs]: Thanks. We probably will need to change logic of embedder to properly select right version of API implementation based on name. At this moment I choose return name, but later we can change `find_wasi_module_name` to return an enum.

**Pull request merged**